### DISCUSSION TITLE
Issues Encountered with Point Markers in Visualization

### DISCUSSION THREAD

**Discussion by smittenx** [foxglove/community#1011] [Q&A] [answered]
Hello,

Firstly, I would like to thank you for your work. It has been very helpful for our tasks. However, we have encountered some issues while using the software. After searching for relevant information, we are still unable to resolve these problems.

1.The point markers we generated sometimes fail to display when the perspective changes (rotation, movement). They reappear when we change to another perspective. Is this due to the large number of point markers we have generated? Is there any solution for this?

2.When we use the 3D camera to 2D camera conversion function, the point markers we created do not display correctly. We can only see the point markers when we zoom out a lot, at which point markers become very, very small, making them invisible in a normal view.

I hope these issues can be resolved. Thank you!

**Comment by achim-k** [comment]
Could you provide a MCAP/rosbag file with your marker messages and detailed reproduction steps?

**Reply by smittenx** [reply]
I have attached the relevant code and files below.

**Comment by smittenx** [comment]
Thank you for your reply. Here is the relevant code and the MCAP file. In the file, you can see that in the 3D camera view, the points can display the point markers regardless of the z-value. However, when switching to the 2D camera, we can see the points when the z-value is very large, but as the z-value decreases, the point markers become sparser and eventually disappear.
[143708_temp.zip](https://github.com/user-attachments/files/16122319/143708_temp.zip)
`visualization_points_publisher_ =
      node_->create_publisher<visualization_msgs::msg::MarkerArray>("/visualization/points", 10);
  visualization_msgs::msg::MarkerArray debug_points_array;
  visualization_msgs::msg::Marker marker_map_;
  marker_map_.header = header_;
  marker_map_.ns = ns;
  marker_map_.id = id;
  marker_map_.type = visualization_msgs::msg::Marker::POINTS;
  marker_map_.action = visualization_msgs::msg::Marker::ADD;
  marker_map_.pose.orientation.w = 1.0;
  marker_map_.scale.x = 0.05;
  marker_map_.scale.y = 0.05;
  marker_map_.points.clear();
  marker_map_.colors.clear();

  geometry_msgs::msg::Point basic_point;
  basic_point.x = rolling_grid.GetPositionByIndex(i, j).first;
  basic_point.y = rolling_grid.GetPositionByIndex(i, j).second;
  basic_point.z = 0.0;

  marker_color.r = static_cast<float>(0) / 255.0;
  marker_color.g = static_cast<float>(0) / 255.0;
  marker_color.b = static_cast<float>(255) / 255.0;

  marker_map_.points.push_back(basic_point);
  marker_map_.colors.push_back(marker_color);

  for (const auto& pair : marker_map) {
    debug_points_array.markers.push_back(pair.second);
  }

  visualization_points_publisher_->publish(debug_points_array);`

![2d](https://github.com/foxglove/community/assets/87402850/1f44d3e2-8ea3-48af-9b11-9d9345f8d58b)
![2d-close](https://github.com/foxglove/community/assets/87402850/aab1449a-74ea-48b8-b1ad-6e0ad4e476b8)
![3d](https://github.com/foxglove/community/assets/87402850/69206bbe-5be0-4019-ad94-ee98d90f8f01)
![3d-close](https://github.com/foxglove/community/assets/87402850/732bbdef-9a44-4c90-8afc-544780518486)

**Reply by achim-k** [reply]
This seems like a potential bug, we will investigate this and keep you updated.

**Reply by smittenx** [reply]
Thank you for looking into this. We appreciate your support and look forward to any updates regarding this issue. If you need any additional information or further details, please let us know.

**Comment by snosenzo** [comment, marked as answer]
@smittenx Thanks for bringing this issue to us! I was able to diagnose the cause of the issue being that the points at a given size within the perspective (3D) camera, are too small to be seen in the orthographic (2D) camera because the orthographic camera does not account for the distance from the camera. To fix this, we will make points a minimum size of 1 pixel when being viewed in 2D mode to keep them visible. 
This issue has been fixed and should be available in the web app soon and it will also be in the next desktop release. 
Let us know if you have any more questions
